src/components/form/HdInputPhone.vue

```diff
@@ -0,0 +1,368 @@
+      selectedCountry: COUNTRY_PHONE_CODES.find((country) => country.code === this.defaultCountry),
+      focusedCountry: null,
+      rule: {
+        validate: (value) => {
+          const hasMoreThan8Digits = value.length >= 8;
+          const hasPlusDigit = value.includes('+');
+          const isValid = hasPlusDigit && hasMoreThan8Digits;
+          return isValid;
+        },
+        errorMessage: 'e.g. +55 555 555 5555',
+      },
+      smallArrowIcon,
+    };
+  },
+  computed: {
+    sortedCountriesList() {
+      // Preferred countries first
+      if (this.preferredCountries.length) {
+        return COUNTRY_PHONE_CODES.sort((a, b) => {
+          if (this.preferredCountries.includes(a.code) && this.preferredCountries.includes(b.code)) {
+            return this.t.COUNTRIES[a.code] > this.t.COUNTRIES[b.code];
+          }
+          if (this.preferredCountries.includes(a.code)) {
+            return -1;
+          } if (this.preferredCountries.includes(b.code)) {
+            return 1;
+          } if (!this.preferredCountries.includes(a.code) && !this.preferredCountries.includes(b.code)) {
+            return (this.t.COUNTRIES[a.code] > this.t.COUNTRIES[b.code]) ? 1 : -1;
+          }
+          return 0;
+        });
+      }
+      return COUNTRY_PHONE_CODES;
+    },
+    arrowClassNames() {
+      const baseClass = 'phone-input__selector__arrow';
+      const directionClass = this.showDropdown ? 'phone-input__selector__arrow--down' : 'phone-input__selector__arrow--up';
+      return [baseClass, directionClass];
+    },
+    t() {
+      return getMessages(this.lang);
+    },
+  },
+  watch: {
+    phoneNumber(newPhoneNumber) {
+      const newSelectedCountry = COUNTRY_PHONE_CODES.find((country) => (
+        country.dial_code === newPhoneNumber.substring(0, MIN_DIGITS_DIAL_CODE)
+        || country.dial_code === newPhoneNumber.substring(0, MIN_DIGITS_DIAL_CODE + 1)
+        || country.dial_code === newPhoneNumber.substring(0, MAX_DIGITS_DIAL_CODE)
+      ));
+      if (newSelectedCountry) this.selectedCountry = newSelectedCountry;
+      action('input')(newPhoneNumber);
+    },
+    selectedCountry(newCountry, prevCountry) {
+      if (this.phoneNumber.length >= MAX_DIGITS_DIAL_CODE) {
+        this.phoneNumber = this.phoneNumber.replace(prevCountry.dial_code, newCountry.dial_code);
+      } else {
+        this.phoneNumber = newCountry.dial_code;
+      }
+    },
+    showDropdown(isShown) {
+      let selectKeys;
+      if (isShown) {
+        selectKeys = window.addEventListener('keyup', this.keysHandler);
+      } else {
+        window.removeEventListener('keyup', selectKeys);
+      }
+    },
+  },
+  methods: {
+    toggleDropdown() {
+      this.showDropdown = !this.showDropdown;
+    },
+    selectCountry(country) {
+      this.selectedCountry = country;
+      this.toggleDropdown();
+      if (this.phoneNumber.length <= MAX_DIGITS_DIAL_CODE) {
+        this.$refs.input.$refs.input.$refs.input.focus();
```

Comment:
Isn't this reference too deep? Are we sure there will always be three-dimensional input?

Reply:
Oh yes, I actually was going to ask if there is a better way to handle this, I feel that it's super insecure doing this, but so far yes, we always are going to have three inputs.

Reply:
hmm.. my idea would be; we can get rid of the `HdInputFormatter` component in this case. I can't see a necessity to use it as we already define the formatting function in the `HdInputPhone.vue` - so, we can remove the `HDInputFormatter` and use the formatting function to format the phone without passing the function through another component. in that way, we can get rid of nested inputs as well. WDYT? does it make sense?

Reply:
I think the usage of `HdInputFormatter` is fine here, or maybe even recommended. It's a simple component that does a simple and specific job. Removing it will require us to replicate its functionality in HdInputPhone, which goes against reusability and modularity.

I think the deep access to the grandchildren is okay here.
A better option would be to expose a method `focusInput` from HdInputFormatter and HdInput, so you can do just `this.$refs.input.focusInput()`, but that's still relaying on the internal API of the component which is not recommended, as it might change in the future without any notice.
An even better option would be to use the exposed API of the component, aka the contract (props and events): We can have a `focused` prop (might even use the native `autofocus`, but I didn't test it), with `false` as the default value, and we can dynamically set it to `true` when we want to focus the component, and of course always set it back to `false` when the input has been blurred.

Reply:
I can try setting the `focus` prop and see how it works, thank you @ilyasmez!
And yes @volcanioo, the reason I used the `HdInputFormatter` in the beginning was to avoid doing the same logic of format the value only when it's not focused